libs/tlv/src/coder.rs

```diff
@@ -0,0 +1,165 @@
+use std::io::{Read, Write};
+
+use crate::length::Length;
+
+/// Encodes a value into a writer.
+pub trait Encode {
+    /// Converts the value into a byte representation and writes it to the
+    /// writer.
+    fn encode<W>(&self, writer: &mut W) -> std::io::Result<()>
+    where
+        W: Write;
+
+    /// Returns the length of the encoded value.
+    fn encoded_length(&self) -> std::io::Result<Length>;
+}
+
+/// Decodes a value from a decoder.
+pub trait Decode
+where
+    Self: Sized,
+{
+    /// Reads the value from the decoder.
+    fn decode<R>(reader: &mut R) -> std::io::Result<(Self, usize)>
+    where
+        R: Read;
+}
+
+macro_rules! impl_coders {
+    ($($t:ty),*) => {
+        $(
+            impl Encode for $t {
+                fn encode<W>(&self, writer: &mut W) -> std::io::Result<()> where W: Write {
+                    writer.write_all(&self.to_be_bytes())
+                }
+
+                fn encoded_length(&self) -> std::io::Result<Length> {
+                    Ok(Length::new(std::mem::size_of::<$t>() as u16))
+                }
+            }
+
+            impl Decode for $t {
+                fn decode<R>(reader: &mut R) -> std::io::Result<(Self, usize)> where R: Read {
+                    const SIZE: usize = std::mem::size_of::<$t>();
+                    let mut buf = [0; SIZE];
+                    reader.read_exact(&mut buf)?;
+                    Ok((<$t>::from_be_bytes(buf), SIZE))
+                }
+            }
+        )*
+    };
+}
+
+impl_coders!(u8, u16, u32, u64, usize, i8, i16, i32, i64, isize);
+
+impl<T: Encode> Encode for &T {
```

Comment:
This is a useful trick for implementing a trait for a whole class of thing. This says "for all `T` that implement `Encode`, here's how to implement `Encode` for `&T`". This means you can just accept `Encode` values and the caller can pass either an owned `T` or a borrowed `&T`:

```rust
let value = Test { a: 99 };
tlv::to_vec(&value); // just borrowing, I want to use it again
tlv::to_vec(value); // okay, I'm done with `value` so pass ownership
```